File: fsharp/diffie-hellman/DiffieHellman.fs
module DiffieHellman

open System
open System.Numerics
open System.Security.Cryptography

let privateKey (primeP: bigint) =
    let rng = new RNGCryptoServiceProvider()
    let bytes = primeP.ToByteArray()
    let mutable r = BigInteger(-1)
    while r < 0I || r >= primeP do
        rng.GetBytes(bytes)
        r <- BigInteger(bytes)
    r

let publicKey primeP primeG privateKey = BigInteger.ModPow(primeG, privateKey, primeP)

let secret primeP publicKey privateKey = BigInteger.ModPow(publicKey, privateKey, primeP)

File: csharp/diffie-hellman/DiffieHellman.cs
using System.Numerics;
using System.Security.Cryptography;

public static class DiffieHellman
{
    public static BigInteger PrivateKey(BigInteger primeP) 
    {
        // https://stackoverflow.com/a/17367241/1595197
        var rng = new RNGCryptoServiceProvider();
        var buffer = primeP.ToByteArray();
        BigInteger result;

        do
        {
            rng.GetBytes(buffer);
            result = new BigInteger(buffer);
        } while (result <= 0 || result >= primeP);

        return result;
    }

    public static BigInteger PublicKey(BigInteger primeP, BigInteger primeG, BigInteger privateKey) =>
        BigInteger.ModPow(primeG, privateKey, primeP);

    public static BigInteger Secret(BigInteger primeP, BigInteger publicKey, BigInteger privateKey) =>
        BigInteger.ModPow(publicKey, privateKey, primeP);
}